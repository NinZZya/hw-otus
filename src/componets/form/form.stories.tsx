import React from "react";
// also exported from "@storybook/react" if you can deal with breaking changes in 6.1
import { Meta } from "@storybook/react/types-6-0";

import { Form } from "./form";
import { Input } from "../inputs/input/input";
import { Button } from "../buttons/button/button";

export default {
  title: "Example/Form",
  component: Form,
} as Meta;

export const EmptyForm: React.FC = () => <Form />;

export const LoginForm: React.FC = () => (
  <Form>
    <Input required type="text" placeholder="login" />
    <Input required type="password" placeholder="password" />
    <Button>Submit</Button>
  </Form>
);
