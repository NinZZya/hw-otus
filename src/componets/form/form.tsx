import React, { FC, FormEvent, ReactNode } from "react";
/** @jsx jsx */
import { jsx, css, useTheme } from "@emotion/react";
import { SerializedStyles } from "@emotion/utils/types";

interface Theme {
  color?: {
    text: string;
    primary: string;
    surface: string;
    shadowBorder: string;
  };
  width: string;
  height: string;
}

type OnSubmit = (evt: FormEvent<HTMLFormElement>) => void;

interface FormProp {
  [key: string]: string | OnSubmit | ReactNode[] | Element[] | undefined;
  onSubmit?: OnSubmit | undefined;
  children?: ReactNode[] | Element[] | undefined;
  width?: string | undefined;
  height?: string | undefined;
}

const DefaultColor = {
  TEXT: "#ffffff",
  SHADOW_BORDER: "rgba(255, 255, 255, 0.1)",
  BACKGROUND: "#3700b3",
  SURFACE: "#121212",
};

const DefaultForm = {
  WIDTH: "400px",
  HEIGHT: "500px",
};

const getStyleForm = (theme: Theme): SerializedStyles => css`
  display: flex;
  flex-direction: column;
  width: ${theme.width};
  height: ${theme.height};
  padding: 3rem;
  color: ${theme.color ? theme.color.text : DefaultColor.TEXT};
  border: 1px solid
    ${theme.color ? theme.color.shadowBorder : DefaultColor.SHADOW_BORDER};
  border-radius: 0.3rem;
  box-sizing: border-box;
  background-color: ${theme.color ? theme.color.surface : DefaultColor.SURFACE};
`;

export const Form: FC<FormProp> = (props) => {
  const theme: Theme = {
    ...useTheme(),
    width: props.width ? props.width : DefaultForm.WIDTH,
    height: props.height ? props.height : DefaultForm.HEIGHT,
  };

  return <form css={getStyleForm(theme)} {...props} />;
};
