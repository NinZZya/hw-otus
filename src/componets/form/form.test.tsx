import React from "react";
import renderer from "react-test-renderer";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Form } from "./form";

configure({ adapter: new Adapter() });

describe(`Form render coorectly`, () => {
  it(`Form without props render coorectly`, () => {
    const tree = renderer.create(<Form />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Form with components render coorectly`, () => {
    const tree = renderer
      .create(
        <Form>
          <input />
          <input />
          <button />
        </Form>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Form was submited`, () => {
    const onSubmit = jest.fn();
    const button = shallow(<Form onSubmit={onSubmit} />);

    button.simulate("submit");
    expect(onSubmit).toHaveBeenCalledTimes(1);
  });
});
