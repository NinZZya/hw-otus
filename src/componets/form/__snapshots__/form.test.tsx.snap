// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form render coorectly Form with components render coorectly 1`] = `
<form
  css={
    Object {
      "map": undefined,
      "name": "1ga4see",
      "next": undefined,
      "styles": "
  display: flex;
  flex-direction: column;
  width: 400px;
  height: 500px;
  padding: 3rem;
  margin: 3rem;
  color: #ffffff;
  border: 1px solid
    rgba(255, 255, 255, 0.1);
  border-radius: 0.3rem;
  box-sizing: border-box;
  background-color: #121212;
",
      "toString": [Function],
    }
  }
>
  <input />
  <input />
  <button />
</form>
`;

exports[`Form render coorectly Form without props render coorectly 1`] = `
<form
  css={
    Object {
      "map": undefined,
      "name": "1ga4see",
      "next": undefined,
      "styles": "
  display: flex;
  flex-direction: column;
  width: 400px;
  height: 500px;
  padding: 3rem;
  margin: 3rem;
  color: #ffffff;
  border: 1px solid
    rgba(255, 255, 255, 0.1);
  border-radius: 0.3rem;
  box-sizing: border-box;
  background-color: #121212;
",
      "toString": [Function],
    }
  }
/>
`;
